tedana/metrics/dependence.py

```diff
@@ -0,0 +1,354 @@
+"""
+Fit models.
+"""
+import logging
+import os.path as op
+
+import numpy as np
+import pandas as pd
+from scipy import stats
+
+from tedana import io, utils
+from tedana.stats import getfbounds, computefeats2, get_coeffs, t_to_z
+
+
+LGR = logging.getLogger(__name__)
+RepLGR = logging.getLogger('REPORT')
+RefLGR = logging.getLogger('REFERENCES')
+
+
+def calculate_weights(data_optcom, mixing_z):
+    """
+    Calculate weights.
+
+    Parameters
+    ----------
+    data_optcom : (S x T) array_like
+        Optimally combined data
+    mixing_z : (T x C) array_like
+        Z-scored mixing matrix
+
+    Returns
+    -------
+    weights : (S x C) array_like
+        Parameter estimates for optimally combined data against the mixing
+        matrix.
+    """
+    # compute un-normalized weight dataset (features)
+    weights = computefeats2(data_optcom, mixing_z, normalize=False)
+    return weights
+
+
+def calculate_betas(data_optcom, mixing):
+    assert data_optcom.shape[1] == mixing.shape[0]
+    # demean optimal combination
+    data_optcom_dm = data_optcom - data_optcom.mean(axis=-1, keepdims=True)
+    # compute PSC dataset - shouldn't have to refit data
+    betas = get_coeffs(data_optcom_dm, mixing)
+    return betas
+
+
+def calculate_psc(data_optcom, optcom_betas):
+    assert data_optcom.shape[1] == optcom_betas.shape[0]
+    PSC = 100 * optcom_betas / data_optcom.mean(axis=-1, keepdims=True)
+    return PSC
+
+
+def compute_countsigFT2(F_T2_clmaps):
+    countsigFT2 = F_T2_clmaps.sum(axis=0)
+    return countsigFT2
+
+
+def compute_countsigFS0(F_S0_clmaps):
+    countsigFS0 = F_S0_clmaps.sum(axis=0)
+    return countsigFS0
+
+
+def compute_countsignal(Z_clmaps):
+    countsignal = Z_clmaps.sum(axis=0)
+    return countsignal
+
+
+def compute_countnoise(Z_maps, Z_cl_maps, z_thresh=1.95):
+    n_components = Z_maps.shape[1]
+    countnoise = np.zeros(n_components)
+    for i_comp in range(n_components):
+        # index voxels significantly loading on component but not from clusters
+        comp_noise_sel = ((np.abs(Z_maps[:, i_comp]) > z_thresh) &
+                          (Z_clmaps[:, i_comp] == 0))
+        countnoise[i_comp] = np.array(comp_noise_sel, dtype=np.int).sum()
```

Comment:
I don't think this needs to be a for loop. From what I can tell this would broadcast without any issue:

```python
comp_noise_sel = (np.abs(Z_maps) > z_thresh) & (Z_clmaps == 0)
countnoise = np.sum(comp_noise_sel).sum(axis=0)
```

Summing the boolean should automatically cast it to int, too. Let me know if I'm misunderstanding the dimensions of `Z_maps` and `Z_clmaps` though!

Reply:
Nice catch! This was originally part of a much larger for loop, so I just copied the loop for all of the smaller functions. This makes things much cleaner.

Reply:
Also, the `comp_noise_sel` and `countnoise` calculation appear in this function and the following two functions. If we consolidate this (removing the for loop, as above), what do you think of explicitly calling this in `compute_signal_minus_noise_z()` and `compute_signal_minus_noise_t()`?

Reply:
I need the noise index in those two more than the counts, so I'll be duplicating most of the work there anyway.